hypothesis-python/src/hypothesis/internal/conjecture/data.py

```diff
@@ -856,31 +856,221 @@ def kill_branch(self) -> None:
         """Mark this part of the tree as not worth re-exploring."""
 
     def draw_integer(
-        self, value: int, *, was_forced: bool, kwargs: IntegerKWargs
+        self, value: int, *, kwargs: IntegerKWargs, was_forced: bool
     ) -> None:
         pass
 
     def draw_float(
-        self, value: float, *, was_forced: bool, kwargs: FloatKWargs
+        self, value: float, *, kwargs: FloatKWargs, was_forced: bool
     ) -> None:
         pass
 
     def draw_string(
-        self, value: str, *, was_forced: bool, kwargs: StringKWargs
+        self, value: str, *, kwargs: StringKWargs, was_forced: bool
     ) -> None:
         pass
 
     def draw_bytes(
-        self, value: bytes, *, was_forced: bool, kwargs: BytesKWargs
+        self, value: bytes, *, kwargs: BytesKWargs, was_forced: bool
     ) -> None:
         pass
 
     def draw_boolean(
-        self, value: bool, *, was_forced: bool, kwargs: BooleanKWargs
+        self, value: bool, *, kwargs: BooleanKWargs, was_forced: bool
     ) -> None:
         pass
 
 
+IRTreeNodeType: TypeAlias = Union["IRTreeNode", "IRTreeLeaf"]
+LocationType: TypeAlias = Tuple[int, int]
+
+
+@attr.s(slots=True)
+class IRTreeLeaf:
+    ir_type: IRTypeName = attr.ib()
+    value: IRType = attr.ib()
+    kwargs: IRKWargsType = attr.ib()
+    was_forced: bool = attr.ib()
+    depth: int = attr.ib()
```

Comment:
Can we end up with very many redundant copies of these objects lying around?  I guess booleans and small ints are all in a small object pool, so it's really just the kwargs...

Reply:
Storing `kwargs` is a necessity for intelligent shrinking, so we don't have a choice but to track all of these, right? Is the concern that we should maybe be pooling these to reference a shared object pool to reduce memory in the case of many ir draws that happen to use the same kwargs?

Reply:
Yep.  We clearly shouldn't do that without a profile showing that it's necessary, but I think it's worth checking a profile; we have some scar-tissue of weird optimizations in `conjecture/` which are in fact important.

Reply:
I'm glad you asked me to check this. A quick check with memray on our quality tests shows a huge increase for e.g. `test_minimize_multiple_elements_in_silly_large_int_range`: 20mb (master) vs 80mb (this pr). I'll confirm this is a result of the tree and work on a pooling implementation.

This probably also affects `DataTree`.

Reply:
kwargs is definitely part of the problem here, but I think actually the worse offender is that we are duplicating the example hierarchy in an unintelligent way in `IRTree`. Existing space-saving optimizations in `Example` are probably relevant.

I'll have to spend some more time thinking about the rigtht forward here. Maybe we can hack the existing `Example` infrastructure to also track our ir kwargs, and construct a more workable `IRTree` definition from that only when moving to shrinking.

Reply:
ok, I'm going to do a bit of a 180 here. Let's move to `Example` to track the ir tree, not as a hack, but as the canonical representation. Forget this `IRTree` stuff. The output of the migrated shrinker is ir leaves, not a tree;tn the tree structure is only for convenience when mutating and is already encoded in `Example` anyway. I'm relatively confident I can make almost all the passes on migrated shrinker work with this approach. "almost" because I haven't thought that deeply about all the passes.

This + pooling solves our memory problems.